server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -333,13 +333,21 @@ public <C extends Collector, T> T search(Query query, CollectorManager<C, T> col
     }
 
     /**
-     * Similar to the lucene implementation but it will wait for all threads to fisinsh before returning even if an error is thrown.
-     * In that case, other exceptions will be ignored and the first exception is thrown after all threads are finished.
+     * Similar to the lucene implementation, with the following changes made:
+     * 1) it will wait for all threads to finish before returning when an exception is thrown. In that case, subsequent exceptions will be
+     * ignored and the first exception is re-thrown after all tasks are completed.
+     * 2) Tasks are cancelled on exception to prevent needless computation
+     * 3) collection is unconditionally offloaded to the executor when set, even when there is a single slice or the request does not
+     * support concurrent collection. The executor is not set only when concurrent search has been explicitly disabled at the cluster level.
      * */
     private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T> collectorManager, C firstCollector) throws IOException {
-        if (queueSizeBasedExecutor == null || leafSlices.length <= 1) {
+        // the executor will be null only when concurrency is disabled at the cluster level
```

Comment:
This comment isn't true and that's a little confusing, we also have no executor if the search doesn't support it.

Reply:
yea the comment was outdated. But is no longer true after the latest updates. There are no exceptions to offloading, so the comment is true again, and we are good :)